src/main/scala/cromwell/engine/workflow/ValidateActor.scala

```diff
@@ -38,44 +49,88 @@ class ValidateActor(wdlSource: WdlSource, workflowInputs: Option[WdlJson], workf
+      validatedWorkflowOptions: Option[WorkflowOptions] <- Future(validateWorkflowOptions(CromwellBackend.backend()))
+      validatedRuntimeAttrs: Seq[Set[String]] <- Future(validateRuntimeAttributes(namespaceWithWorkflow))
+    } yield ValidationSuccess(namespaceWithWorkflow, validatedInputs, validatedWorkflowOptions, validatedRuntimeAttrs)) recover {
+      case reason: Throwable => ValidationFailure(reason)
+    }
+
+    //TODO: IMO, the responsibility for returning a status code should be a part of the client. This class should only return a validation result,
+    // and the callee should convert that to a status code or anything that it wants
 
     // Now validate that this Future completed successfully:
-    futureValidation onComplete {
-      case Success(_) =>
+    futureValidation map {
+      case _: ValidationSuccess =>
         logger.info(s"$tag success $sentBy")
         sentBy ! RequestComplete(
           StatusCodes.OK,
           APIResponse.success("Validation succeeded."))
 
-      case Failure(ex) =>
-        val messageOrBlank = Option(ex.getMessage).mkString
+      case ValidationFailure(reason) =>
+        val messageOrBlank = Option(reason.getMessage).mkString
         logger.info(s"$tag error $sentBy: $messageOrBlank")
         sentBy ! RequestComplete(
           StatusCodes.BadRequest,
-          APIResponse.fail(ex))
+          APIResponse.fail(reason))
+    }
+  }
+
+  //TODO: With PBE, this should be defined in the backend.
+  private def validateRuntimeAttributes(namespaceWithWorkflow: NamespaceWithWorkflow): Seq[Set[String]] = {
+    TryUtil.sequence(namespaceWithWorkflow.workflow.calls map {
+      call => CromwellRuntimeAttributes.validateKeys(call.task.runtimeAttributes.attrs.keySet, CromwellBackend.backend().backendType)
+    }) match {
+      case Success(validatedRuntimeAttrs) => validatedRuntimeAttrs
+      case Failure(reason) => throw new IllegalArgumentException("Failed to validate runtime attributes.", reason)
+    }
+  }
+
+  private def validateInputs(namespaceWithWorkflow: NamespaceWithWorkflow): Option[WorkflowCoercedInputs] = {
+    workflowInputs match {
+      case Some(json) =>
+        def rawInputs(): Map[String, JsValue] = {
+          Try(json.parseJson) match {
+            case Success(JsObject(inputs)) => inputs
+            case Failure(reason: Throwable) => throw new Throwable(reason.getMessage)
+            case _ => throw new IllegalArgumentException(s"Failed to parse the workflow inputs JSON: $workflowInputs")
+          }
+        }
+
+        namespaceWithWorkflow.coerceRawInputs(rawInputs()) match {
+          case Success(coercedInputs) => Option(coercedInputs)
+          case Failure(reason: ThrowableWithErrors) => throw new ValidationException(reason.message, reason.errors)
+          case Failure(reason) => throw new IllegalArgumentException("Failed to coerce inputs", reason)
```

Comment:
Is this catching above  case Failure(reason: Throwable) => throw new Throwable(reason.getMessage)  ?